internal/parser/parser.go

```diff
@@ -210,33 +209,68 @@ func (p *parser) parseArguments(elementKind string) (*token, *ast.Arguments, boo
 }
 
 // TODO(sbarzowski) - this returned bool is weird
-func (p *parser) parseParameters(elementKind string) (*token, *ast.Parameters, bool, error) {
-	parenR, args, trailingComma, err := p.parseArguments(elementKind)
-	if err != nil {
-		return nil, nil, false, err
-	}
-	var params ast.Parameters
-	for _, arg := range args.Positional {
-		idFodder, id, ok := astVarToIdentifier(arg.Expr)
-		if !ok {
-			return nil, nil, false, errors.MakeStaticError(fmt.Sprintf("Expected simple identifier but got a complex expression."), *arg.Expr.Loc())
-		}
-		params.Required = append(params.Required, ast.CommaSeparatedID{
-			NameFodder:  idFodder,
-			Name:        *id,
-			CommaFodder: arg.CommaFodder,
-		})
-	}
-	for _, arg := range args.Named {
-		params.Optional = append(params.Optional, ast.NamedParameter{
-			NameFodder:  arg.NameFodder,
-			Name:        arg.Name,
-			EqFodder:    arg.EqFodder,
-			DefaultArg:  arg.Arg,
-			CommaFodder: arg.CommaFodder,
-		})
+func (p *parser) parseParameters(elementKind string) (*token, []ast.Parameter, bool, error) {
+
+	var parenR *token
+	var params []ast.Parameter
+	gotComma := false
+	commaFodder := ast.Fodder{}
+	first := true
+	for {
+		next := p.peek()
+
+		if next.kind == tokenParenR {
+			// gotComma can be true or false here.
+			parenR = p.pop()
+			break
+		}
+
+		if !first && !gotComma {
+			return nil, nil, false, errors.MakeStaticError(fmt.Sprintf("Expected a comma before next %s, got %s.", elementKind, next), next.loc)
+		}
+
+		idFodder, id, eqFodder, expr, err := p.parseArgument()
+		if err != nil {
+			return nil, nil, false, err
+		}
+
+		if p.peek().kind == tokenComma {
+			comma := p.pop()
+			gotComma = true
+			commaFodder = comma.fodder
+		} else {
+			gotComma = false
+		}
+
+		if id == nil {
+			// Param has no default, is just 'id2'
+			idFodder2, id2, ok := astVarToIdentifier(expr)
```

Comment:
Since we have separated this from `parseArguments` anyway, maybe we can get rid of this hack? I mean, let's create a new function `parseParameter` which treats a single identifier as parameter name, not as a variable reference. I think we'll be able to make the code here much nicer then (e.g. we won't need `id == nil` condition at all, we'll be able to refactor parseArgument/parseParameter to return the argument/parameter struct instead of its composite parts).